### ISSUE TITLE
Can a node inherit a property of the parent?

### ISSUE THREAD

**Issue by JonesKenzie** [django-mptt/django-mptt#770]
Is there a way to ensure that a when creating a tree all nodes will inherit a field or property of the root? For example the root has a field called Color that was set to "red". How can I force that every node I create have the same "red" in Color?

**Comment by timoludwig** [comment]
This has nothing to do with django-mptt, it is a pure Python/Django question.
You could solve this e.g. with a [property](https://docs.python.org/3/library/functions.html#property):
```
class MyClass(MPTTModel):
    parent = TreeForeignKey(
        "self", blank=True, null=True, on_delete=models.PROTECT
    )
    _color = models.CharField(max_length=255, blank=True)

    @property
    def color(self):
        if self.parent:
            return self.parent._color
        return self._color

    @color.setter
    def color(self, value):
        self._color = value
```

**Comment by matthiask** [comment]
You could also copy the `color` value from the `parent` in an overridden save method. However, there's nothing stopping anyone from circumventing `.save()`, e.g. by going directly through `.update()`

Another possibility is to use a recursive CTE to fetch the tree or portions of it, and ensure that the color value is only ever selected from root nodes. Just as an example, insert `color as tree_color,` in the `SELECT` before the `UNION ALL` and `__tree.tree_color` in the `SELECT` after the `UNION ALL`. https://github.com/matthiask/django-tree-queries/blob/219c14ea1cf2b3d6a1bf793ed159a3bbd455df13/tree_queries/compiler.py#L51

There are many ways to do what you want but the preferrable solution depends on your constraints.

**Comment by JonesKenzie** [comment]
> This has nothing to do with django-mptt, it is a pure Python/Django question.
> You could solve this e.g. with a [property](https://docs.python.org/3/library/functions.html#property):
> 
> ```
> class MyClass(MPTTModel):
>     parent = TreeForeignKey(
>         "self", blank=True, null=True, on_delete=models.PROTECT
>     )
>     _color = models.CharField(max_length=255, blank=True)
> 
>     @property
>     def color(self):
>         if self.parent:
>             return self.parent._color
>         return self._color
> 
>     @color.setter
>     def color(self, value):
>         self._color = value
> ```

Yes, I realized this when doing my implementation. I've opted to override the save method, but maybe I should also do this to improve the enforcement. Thanks @timoludwig !

@matthiask I would prefer not to mess up with the CTEs, but in this case I would need to create a trigger, isn't that right? A major thing is to automatically update the nodes when the root is changed. 

Also, when implementing forms, to block the field from appearing for non-root nodes I tried accessing the instance as suggest in a Stack OverFlow post, but it doesn't work as expected. 

```
 from django import forms
from .models import ProductTemplate
class ProductAdminForm(forms.ModelForm):
   
    class Meta:
        model = ProductTemplate
        fields = '__all__'
    def __init__(self, *args, **kwargs):
        super(ProductAdminForm, self).__init__(*args, **kwargs)
        
        instance = getattr(self, 'instance', None)

        if instance and instance.is_root_node():
            self.fields["color"].widget.attrs['readonly'] = True
                ```

**Comment by matthiask** [comment]
@JonesKenzie With CTEs there wouldn't be a requirement to keep the data in sync at all (using triggers, save handlers or whatnot), because the inheritance of property values happens at `SELECT` time. I certainly understand if you do not want to touch them :-) but – as a heavy user of django-mptt – I was surprised by the finding that CTEs perform well enough for the trees which I have in the database.

Regarding removing fields from forms: `if not instance or not instance.is_root_node(): self.fields.pop("color")` would maybe work too. Or disabling the field. I wouldn't recommend setting the `readonly` attribute, you also want to disable the backend-part of the processing.

**Comment by JonesKenzie** [comment]
@matthiask  It doesn't work :( The thing is the forms don't give the instance.

**Comment by matthiask** [comment]
See #833 — the project is in maintenance-only mode. For usage questions like this one, Stack Overflow or the Django Forum are usually better venues. Thanks for your understanding.
